pkg/networkservice/common/connect/server.go

```diff
@@ -94,60 +125,67 @@ func (s *connectServer) Close(ctx context.Context, conn *networkservice.Connecti
-
 	clientURL := clienturlctx.ClientURL(ctx)
-	if clientURL == nil {
-		err := errors.Errorf("clientURL not found for incoming connection: %+v", conn)
-		return injecterror.NewClient(err)
-	}
 
 	// First check if we have already requested some clientURL with this conn.GetID().
 	if connInfo, ok := s.connInfos.Load(conn.GetId()); ok {
 		if *connInfo.clientURL == *clientURL {
 			return connInfo.client
 		}
-		// For some reason we have changed the clientURL, so we need to close the existing client.
-		if _, clientErr := connInfo.client.Close(ctx, conn); clientErr != nil {
+
+		// For some reason we have changed the clientURL, so we need to close and delete the existing client.
+		if _, clientErr := connInfo.client.client.Close(ctx, conn); clientErr != nil {
 			logger.Warnf("failed to close client: %+v", clientErr)
 		}
-	}
 
-	// Fast path if we already have client for the clientURL, use it.
-	client, loaded := s.clients.Load(clientURL.String())
-	if !loaded {
-		// If not, create and LoadOrStore a new one.
-		newClient, cancel := s.newClient(clientURL)
-		client, loaded = s.clients.LoadOrStore(clientURL.String(), newClient)
-		if loaded {
-			// No one will use `newClient`, it should be canceled.
-			cancel()
-		}
+		s.closeClient(connInfo.client, connInfo.clientURL.String())
 	}
 
-	s.connInfos.Store(conn.GetId(), connectionInfo{
-		clientURL: clientURL,
-		client:    client,
+	var client *clientInfo
+	<-s.executor.AsyncExec(clientURL.String(), func() {
+		// Fast path if we already have client for the clientURL and we chould not reconnect, use it.
+		var loaded bool
+		client, loaded = s.clients.Load(clientURL.String())
+		if !loaded {
+			// If not, create and LoadOrStore a new one.
+			client = s.newClient(clientURL)
+			s.clients.Store(clientURL.String(), client)
+		}
+		client.count++
 	})
 
 	return client
 }
 
-func (s *connectServer) newClient(clientURL *url.URL) (networkservice.NetworkServiceClient, context.CancelFunc) {
-	clientPtr := new(networkservice.NetworkServiceClient)
-
+func (s *connectServer) newClient(clientURL *url.URL) *clientInfo {
 	ctx, cancel := context.WithCancel(s.ctx)
-	onClose := func() {
-		if deleted := s.clients.Delete(clientURL.String()); deleted {
-			cancel()
-		}
+	client := clienturl.NewClient(clienturlctx.WithClientURL(ctx, clientURL), s.clientFactory, s.clientDialOptions...)
+	return &clientInfo{
+		client:  client,
+		count:   0,
+		onClose: cancel,
 	}
+}
 
-	*clientPtr = chain.NewNetworkServiceClient(
-		&connectClient{onClose: onClose},
-		clienturl.NewClient(clienturlctx.WithClientURL(ctx, clientURL), s.clientFactory, s.clientDialOptions...),
-	)
+func (s *connectServer) closeClient(client *clientInfo, clientURL string) {
+	<-s.executor.AsyncExec(clientURL, func() {
+		client.count--
+		if client.count == 0 {
+			if loadedClient, ok := s.clients.Load(clientURL); ok && client == loadedClient {
+				s.clients.Delete(clientURL)
+			}
+			client.onClose()
+		}
+	})
```

Comment:
Could be made non-blocking with something like:

```suggestion
	if atomic.AddInt64(&client.count,-1) == 0 {
		client,ok := s.clients.LoadAndDelete(clientURL)
		if ok {
			client.onClose()
		}
	}
```

Reply:
Ah... I'd missed that this was the multiexecutor and thus per clientURL and not global. This is fine as written.